Bio/Align/__init__.py

```diff
@@ -932,6 +933,82 @@ def sort(self, key=None, reverse=False):
         else:
             self._records.sort(key=key, reverse=reverse)
 
+    @property
+    def substitutions(self):
+        """Return an Array with the number of substitutions of letters in the alignment.
+
+        As an example, consider a multiple sequence alignment of three DNA sequences:
+
+        >>> from Bio.Seq import Seq
+        >>> from Bio.SeqRecord import SeqRecord
+        >>> from Bio.Align import MultipleSeqAlignment
+        >>> seq1 = SeqRecord(Seq("ACGT"), id="seq1")
+        >>> seq2 = SeqRecord(Seq("A--A"), id="seq2")
+        >>> seq3 = SeqRecord(Seq("ACGT"), id="seq3")
+        >>> seq4 = SeqRecord(Seq("TTTC"), id="seq4")
+        >>> alignment = MultipleSeqAlignment([seq1, seq2, seq3, seq4])
+        >>> print(alignment)
+        Alignment with 4 rows and 4 columns
+        ACGT seq1
+        A--A seq2
+        ACGT seq3
+        TTTC seq4
+
+        >>> m = alignment.substitutions
+        >>> print(m)
+            A   C   G   T
+        A 3.0 0.5 0.0 2.5
+        C 0.5 1.0 0.0 2.0
+        G 0.0 0.0 1.0 1.0
+        T 2.5 2.0 1.0 1.0
+        <BLANKLINE>
+
+        Note that the matrix is symmetric, with counts divided equally on both
+        sides of the diagonal. For example, the total number of substitutions
+        between A and T in the alignment is 3.5 + 3.5 = 7.
+
+        Any weights associated with the sequences are taken into account when
+        calculating the substitution matrix.  For example, given the following
+        multiple sequence alignment::
+
+            GTATC  0.5
+            AT--C  0.8
+            CTGTC  1.0
+
+        For the first column we have::
+
+            ('A', 'G') : 0.5 * 0.8 = 0.4
+            ('C', 'G') : 0.5 * 1.0 = 0.5
+            ('A', 'C') : 0.8 * 1.0 = 0.8
+
+        """
```

Comment:
Is it helpful to spell out what happens with the second column - i.e. how are gaps treated?

Reply:
Substitution matrices don't include gaps by textbook definition.

Reply:
That isn't obvious to me - I'd prefer a brief mention of this in the docstring.

Reply:
Think of a BLOSUM matrix - it does not include gaps. In an alignment, gaps are assessed by gap open / extend penalties, not by a substitution matrix. It would not even be possible to include gaps in a substitution matrix in a meaningful way. For example, the expected frequency of gaps is not defined.

Reply:
Finding a good place to add that line to this docstring would be great: " In an alignment, gaps are assessed by gap open / extend penalties, not by a substitution matrix."

Reply:
Sorry but I really want to leave this to the textbooks. It's too easy to misinterpret this. We can put something in the tutorial, but as this is the docstring for the `substitutions`, it should focus on what `substitutions` does; gaps are not substitutions.

Reply:
Then how about adding "Gaps are not considered for substitutions"? But as you wish.

Reply:
It's not just that gaps are not considered for substitutions; gaps and substitutions are different things altogether. They may just look similar when we print an alignment and show gaps as dashes.